# HG changeset patch
# User Rik <rik@octave.org>
# Date 1532317202 25200
#      Sun Jul 22 20:40:02 2018 -0700
# Node ID db326f3aacf46c35b5ca4877ec63fa355d8a5dd4
# Parent  fbc2730a653b3afe4186c05bc98b6cec2a30b641
Improve fread speed by a further 2.5X when SKIP parameter used (bug #54100).

* oct-stream.cc (stream::fread): Use C++ seekg and tellg versus C seek and tell
for a 2.5X speedup.

diff --git a/libinterp/corefcn/oct-stream.cc b/libinterp/corefcn/oct-stream.cc
--- a/libinterp/corefcn/oct-stream.cc
+++ b/libinterp/corefcn/oct-stream.cc
@@ -6604,20 +6604,20 @@ namespace octave
     else
       {
         std::istream& is = *isp;
 
         // Initialize eof_pos variable just once per function call
         off_t eof_pos = 0;
         if (skip != 0 && is && ! is.eof ())
           {
-            off_t orig_pos = tell ();
-            seek (0, SEEK_END);
-            eof_pos = tell ();
-            seek (orig_pos, SEEK_SET);
+            off_t orig_pos = is.tellg ();
+            is.seekg (0, is.end);
+            eof_pos = is.tellg ();
+            is.seekg (orig_pos, is.beg);
           }
 
         std::list <void *> input_buf_list;
 
         while (is && ! is.eof ()
                && (read_to_eof || tmp_count < elts_to_read))
           {
             if (! read_to_eof)
@@ -6641,23 +6641,23 @@ namespace octave
             tmp_count += nel;
 
             input_buf_list.push_back (input_buf);
 
             if (is && skip != 0 && nel == block_size)
               {
                 // Attempt to skip.
                 // If skip would move past EOF, position at EOF.
-                off_t orig_pos = tell ();
+                off_t orig_pos = is.tellg ();
                 off_t remaining = eof_pos - orig_pos;
 
                 if (remaining < skip)
-                  seek (0, SEEK_END);
+                  is.seekg (0, is.end);
                 else
-                  seek (skip, SEEK_CUR);
+                  is.seekg (skip, is.cur);
 
                 if (! is)
                   break;
               }
           }
 
         if (read_to_eof)
           {
